ctapipe/io/tableloader.py

```diff
@@ -0,0 +1,331 @@
+                keys="optics_index",
+                # conflicts for TAB_VER, TAB_TYPE, not needed here, ignore
+                metadata_conflicts="silent",
+            )
+
+            table.remove_columns(["optics_index", "camera_index"])
+            self.instrument_table = table
+
+        self.shower_table = None
+        if self.load_simulated and SHOWER_TABLE in self.h5file:
+            self.shower_table = read_table(self.h5file, SHOWER_TABLE)
+
+    def close(self):
+        self.h5file.close()
+
+    def __del__(self):
+        self.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    def _get_reconstructors(self):
+
+        if GEOMETRY_GROUP in self.h5file:
+            g = self.h5file.root[GEOMETRY_GROUP]
+        else:
+            raise ValueError(f"No DL2 geometry data in h5file: {self.h5file}")
+        
+        reconstructors = g._v_children.keys()
+        
+        return reconstructors
+
+    def _read_telescope_table(self, group, tel_id):
+
+        if self.structure == "by_id":
+            key = f"{group}/tel_{tel_id:03d}"
+            condition = None
+        else:
+            key = f"{group}/{self.subarray.tel[tel_id]!s}"
+            condition = f"tel_id == {tel_id}"
+
+        if key in self.h5file:
+            table = read_table(self.h5file, key, condition=condition)
+        else:
+            table = Table()
+
+        return table
+    
+    def _read_telescope_DL2_geometry(self, group, reconstructor):
+        
+        key = f"{group}/{reconstructor}"
+
+        if key in self.h5file:
+            table = read_table(self.h5file, key, condition=None)
+        else:
+            table = Table()
+
+        return table
+
+    def read_data(self, labels=None, sort_by=["obs_id", "event_id"]):
+        """Read data from file.
+        
+        Parameters
+        ----------
+        labels: list
+            Either tel_ids or tel_types depending on dataset split.
+        sort_by: list
+            Columns to sort the final table.
+            By default it is sorted by obs_id and event_id.
+        
+        Returns
+        -------
+        output: astropy.Table
+            Single table containing the enabled data information."""
+        
+        if not any(self.all_loads):
+            self.log.critical("No data enabled for loading.")
+            self.close()
+            return None # it should probably return some exit status != 0
+
+        if (sum(self.all_loads)>1) and self.load_instrument and (not any([self.load_dl1_images, self.load_dl1_parameters, self.load_true_images])):
+            self.log.critical("""Instrument data requires tel_id-based quantities, 
+            but only event_id-based quantities have been selected together with it.""")
+            self.log.critical("No data enabled for loading.")
+            self.close()
+            return None # it should probably return some exit status != 0
```

Comment:
This is not a command line tool. Raise exception on errors. Don't use logging or return codes.

Reply:
Ok for the raises, but why not logging?

Also, what about ``def __exit__(self, exc_type, exc_value, traceback)``?
We don't seem to use the arguments, we just close the file, should we modify/use this or simplify it as only ``def __exit__(self)``?

Reply:
done

Reply:
>  Also, what about def __exit__(self, exc_type, exc_value, traceback)?

Context manager expect those arguments as part of their protocol, even if you don't use them.  I don't think you are free to remove them

Reply:
> Ok for the raises, but why not logging?

Because that's double communication. Logging should happen when desired by the code *catching* the exception, not the code *raising* it. We don't do logging for raised exceptions anywhere else.

Reply:
> def __exit__(self, exc_type, exc_value, traceback)

This is required by the python data model. You cannot change the signature of the magic methods, since python will call `__exit__` with those arguments, even when they are not used inside:

```
class BadContext:
    def __enter__(self):
        return self
    def __exit__(self):
        pass

with BadContext():
    pass
```

Results in:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-cba4f71839bc> in <module>
      6 
      7 with BadContext():
----> 8     pass
      9 

TypeError: __exit__() takes 1 positional argument but 4 were given```

Reply:
this should be OK now (I left __exit__)